core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -191,6 +191,14 @@ class KafkaApis(val requestChannel: RequestChannel,
         s"${leaderAndIsrRequest.brokerEpoch()} smaller than the current broker epoch ${controller.brokerEpoch}")
       sendResponseExemptThrottle(request, leaderAndIsrRequest.getErrorResponse(0, Errors.STALE_BROKER_EPOCH.exception))
     } else {
+      leaderAndIsrRequest.partitionStates().keySet().asScala.find(tp => tp.topic == GROUP_METADATA_TOPIC_NAME).foreach { _ =>
```

Comment:
I think it might be more natural to do this when receiving an `UpdateMetadata` request. Not all brokers will receive `LeaderAndIsr`.

Also, rather than reloading the number of partitions from zk on every LeaderAndIsr change, perhaps we could reload only if we receive a metadata update for a partition that we didn't expect to exist. So if we think there are 10 partitions in __consumer_offsets, we do not reload after an update to partition 0, but we would if we see an update for partition 12.